# **Decision Tree Algorithm**

**What is a Decision Tree?**
> A decision tree is a flowchart-like tree structure where an internal node represents a feature(or attribute), the branch represents a decision rule, and each leaf node represents the outcome. The topmost node in a decision tree is known as the root node. It learns to partition on the basis of the attribute value.
**Lay man example of Decision Tree:**
> Suppose you want to predict whether a person is fit or unfit, given their information like age, eating habits, and physical activity. The decision tree algorithm will start by finding the root node (the best feature to split the dataset). The root node is then split into decision nodes, which are further split into leaf nodes. The leaf node is the prediction of the decision tree.

**Classification Tree (`Categorical Output`)**: Decision Tree which is used to solve classification problems is called a Classification Tree. In a classification tree, we try to use the tree to classify the input into a category. For example, if we want to classify whether a person is fit or unfit, the decision tree will classify the person as fit or unfit based on input variables.

**Regression Tree (`Continuous Output`)**: Decision Tree which is used to solve regression problems is called a Regression Tree. In a regression tree, we try to use the tree to predict the output value. For example, if we want to predict the sales of a company, the decision tree will predict the sales based on input variables.

**Image for Classficication and Regression Tree:**

![image]("C:/Users/Hp/Desktop/Data-Science-Codanics/07_machine_learning/decision_tree_for_heart_attack_prevention_2140bd762d.png")

### **Elements of Decision Tree:**

**Node**:  Point where the data is split. Node in a decision tree represents an attribute. There are two types of nodes:
- `Decision Node`: When a sub-node splits into further sub-nodes, then it is called the decision node.
    - It gives you final decision based on the input features.
    - Final imput might be numeric (regression) or categorical (classofication).
- `Leaf/ Terminal Node`: Nodes do not split is called Leaf or Terminal node.

**Root Node**: The top node where the data is split is called the root node. It represents the entire population or sample and this further gets divided into two or more homogeneous sets. 
- maximum impurity at this level. 

**Branch**: A sub-section of the entire tree is called the branch or sub-tree.

**Splitting**: It is a process of dividing a node into two or more sub-nodes.

**Pruning**: When we remove sub-nodes of a decision node, this process is called pruning. You can say the opposite process of splitting.It is helpful to reduce overfitting.

### **Entropy, Gini Impurity and Information Gain Theory:**

#### **Entropy**: 
> Measure of randomness or uncertainty. It is used to measure the impurity or randomness in the data. 
- It is used to decide how a decision tree can split the data. 
- It is a measure of impurity in a bunch of examples. 
- If entropy is zero then all elements are of the same class, indicating no disorder
- If entropy is one when all elements are of different classes, indicating maximum disorder.

The `formula` for entropy is:

$$
Entropy = -\sum_{i=1}^{n} p_i \cdot \log_2(p_i)
$$    
    Where, $p_i$ is the probability of finding an element of class i.

#### **Information Gain**: 
> It measured the reduction in entropy or impurity in the target variable brought about by partitioning the data based on a certain attribute. 
- It is used in decision trees to decide the order of attribute nodes.
- It is the difference between the entropy of the dataset before the split and after the split.
- `A higher information gain means a higher reduction in entropy, suggesting a better split.`

The `formula` for Information Gain is:

$$
Information\ Gain = Entropy(Parent) - \sum_{i=1}^{n} \left( \frac{|D_v|}{|D|} \cdot Entropy(Child_i) \right)
$$

where:
- $Entropy(Parent)$ is the entropy of the parent node,
- $Entropy(Child_i)$ is the entropy of the $i$-th child node,
- $|D_v|$ is the number of instances that reach child node $i$,
- $|D|$ is the total number of instances at the parent node,
- The sum is over all child nodes of the parent node.


The information gain is based on the decrease in entropy after a dataset is split on an attribute. Constructing a decision tree is all about finding attribute that returns the highest information gain (i.e., the most homogeneous branches).

#### **Gini Impurity**: 
> Gini impurity is a measure of how often a randomly chosen element from the set would be incorrectly labeled if it was randomly labeled according to the distribution of labels in the subset. 

- Gini impurity is zero when all the elements are of a certain class, indicating purity.
- As the distribution of classes approaches uniform, Gini impurity approaches its maximum value.

The `formula` for Gini Impurity is:
$$
Gini\ Impurity (S) = 1 - \sum_{i=1}^{n} p_i^2
$$
Where, $p_i$ is the probability of finding an element of class i.

### **Comparison between Entropy and Gini Impurity:**

| Criteria | Entropy | Gini Impurity |
| --- | --- | --- |
| **Definition** | Entropy is a measure of disorder or uncertainty. It is used to calculate the homogeneity of the input set. | Gini Impurity is a measure of misclassification, which applies in a multi-class classifier context. |
| **Computation** | Computationally, entropy is more intensive as it involves logarithmic calculations. | Gini Impurity is less computationally intensive as it involves only squaring and addition/subtraction. |
| **Sensitivity** | Entropy is more sensitive towards changes in the node probabilities. | Gini Impurity is less sensitive to changes in the node probabilities. |
| **Bias** | Entropy might be slightly biased towards multi-level attributes. | Gini Impurity is unbiased as it tends to isolate the most frequent class in its own branch of the tree. |

### **Summary:**
The **Entropy** and **Information Gain** method `focuses on purity and impurity in a node`. The **Gini** Index or Impurity `measures the probability for a random instance being misclassified when chosen randomly`. The lower the Gini Index, the better the lower the likelihood of misclassification.

#### **1. Import Libarary:**

In [4]:
import math

#### **2. Sample Dataset:**
- Let's say we have a dataset with two classes, A and B and A has 4 elements and B has 6 elements.

In [5]:
# Sample Data for the two classes

n_A = 4  # Number of elements in class A
n_B = 6  # Number of elements in class B
total = n_A + n_B

#### **3. Let's Calculate the Proportion:**
- The proportion of A is 4/10 and the proportion of B is 6/10.

In [6]:
# let's calculate the proportions
p_A = n_A / total  # 4/10 = 0.4
p_B = n_B / total  # 6/10 = 0.6

# print the proportions
print("Proportion of A: ", p_A)
print("Proportion of B: ", p_B)

Proportion of A:  0.4
Proportion of B:  0.6


# **Let's Undertsand the important Metrics of Decision Tree Algorithm.**
1. **Entropy**
2. **Gini Impurity**
3. **Information Gain** 

#### **4. Let's Calculate the Entropy:**
- The entropy of the dataset is calculated as:
- $Entropy = -\sum_{i=1}^{n} p_i \cdot \log_2(p_i)$
- $Entropy = -\left( \frac{4}{10} \cdot \log_2\left(\frac{4}{10}\right) + \frac{6}{10} \cdot \log_2\left(\frac{6}{10}\right) \right)$
- $Entropy = -\left( \frac{4}{10} \cdot \log_2\left(\frac{4}{10}\right) + \frac{6}{10} \cdot \log_2\left(\frac{6}{10}\right) \right)$
- $Entropy = -\left( 0.4 \cdot \log_2(0.4) + 0.6 \cdot \log_2(0.6) \right)$
- $Entropy = -\left( 0.4 \cdot (-0.528) + 0.6 \cdot (-0.442) \right)$
- $Entropy = -\left( -0.2112 - 0.2652 \right)$
- $Entropy = -\left( -0.4764 \right)$
- $Entropy = 0.4764$
- The entropy of the dataset is 0.97.

In [7]:
# Calculate the Entropy
# Entropy is a measure of uncertainty
entropy = -p_A * math.log2(p_A) - p_B * math.log2(p_B)
print("Entropy: ", entropy)

Entropy:  0.9709505944546686


- `Entropy = 0.97` shows that the dataset is `highly impure`. 
- Entropy = 0 means the dataset is `pure` and Entropy = 1 means the dataset is `completely impure`.

#### **5. Let's Calculate the Gini Impurity:**
- The Gini impurity of the dataset is calculated as:
- $Gini\ Impurity (S) = 1 - \sum_{i=1}^{n} p_i^2$
- $Gini\ Impurity (S) = 1 - \left( \left(\frac{4}{10}\right)^2 + \left(\frac{6}{10}\right)^2 \right)$
- $Gini\ Impurity (S) = 1 - \left( \left(\frac{4}{10}\right)^2 + \left(\frac{6}{10}\right)^2 \right)$
- $Gini\ Impurity (S) = 1 - \left( \left(\frac{4}{10}\right)^2 + \left(\frac{6}{10}\right)^2 \right)$
- $Gini\ Impurity (S) = 1 - \left( \left(\frac{16}{100}\right) + \left(\frac{36}{100}\right) \right)$
- $Gini\ Impurity (S) = 1 - \left( 0.16 + 0.36 \right)$
- $Gini\ Impurity (S) = 1 - 0.52$
- $Gini\ Impurity (S) = 0.48$
- The Gini impurity of the dataset is 0.48.

In [8]:
# gini impurity
# Gini impurity is a measure of misclassification
gini = 1- p_A**2 - p_B**2
print("Gini Impurity: ", gini)

Gini Impurity:  0.48


- `Gini Impurity = 0.48` shows that the dataset is `moderate impure`.
- Gini Impurity = 0 means the dataset is `pure` and Gini Impurity = 1 means the dataset is `completely impure`.

#### **6. Let's Calculate the Information Gain:**
- The information gain is the difference between the entropy of the dataset before the split and the sum of the entropy of each group after the split.
- $Information\ Gain = Entropy(Parent) - \sum_{i=1}^{n} \left( \frac{|D_v|}{|D|} \cdot Entropy(Child_i) \right)$
- $Information\ Gain = 0.97 - \left( \frac{4}{10} \cdot 0.811 + \frac{6}{10} \cdot 0.971 \right)$
- $Information\ Gain = 0.97 - \left( 0.3244 + 0.5826 \right)$
- $Information\ Gain = 0.97 - 0.907$
- $Information\ Gain = 0.063$
- The information gain of the dataset is 0.063.

In [9]:
# Information Gain
# Assuming a split on some feature divides the dataset into two subsets
# Subset 1: 2 elements of A, 3 of B
# Subset 2: 2 elements of A, 3 of B
# Entropy and size for each subset
n_1_A, n_1_B = 2, 3
n_2_A, n_2_B = 2, 3

p_1_A = n_1_A / (n_1_A + n_1_B)
p_1_B = n_1_B / (n_1_A + n_1_B)
entropy_1 = -p_1_A * math.log2(p_1_A) - p_1_B * math.log2(p_1_B) if p_1_A and p_1_B else 0

p_2_A = n_2_A / (n_2_A + n_2_B)
p_2_B = n_2_B / (n_2_A + n_2_B)
entropy_2 = -p_2_A * math.log2(p_2_A) - p_2_B * math.log2(p_2_B) if p_2_A and p_2_B else 0

# Calculating information gain
info_gain = entropy - ((n_1_A + n_1_B) / total * entropy_1 + (n_2_A + n_2_B) / total * entropy_2)
print("Information Gain: ", info_gain)

Information Gain:  0.0


`Information Gain = 0.063` shows that the dataset is `highly impure`.
- The information gain from the chosen split is 0.0. This result implies that the split did not reduce the entropy or disorder of the dataset. In other words, the split did not add any additional information that could help distinguish between classes A and B more effectively than before.

These metrics provide insight into the nature of the dataset and the effectiveness of potential splits when constructing a decision tree. In practical applications, you would use these calculations to choose the best feature and split at each node in the tree to maximize the purity of the subsets created.

---
# **Decision Tree `Classification` Example in Python**

#### **1. Import Libraries:**

In [10]:
# import libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import LabelEncoder
from sklearn.impute import SimpleImputer

#### **2. Load the 'Titanic' dataset from Seaborn Library:**

In [11]:
# load the dataset
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

#### **3. Data Preprocessing:**

#### **3.1 Checking for Missing Values:**

In [12]:
# check for missing values
df.isnull().sum().sort_values(ascending=False)

deck           688
age            177
embarked         2
embark_town      2
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
class            0
who              0
adult_male       0
alive            0
alone            0
dtype: int64

- From the above output, we can see that the columns 'deck', 'age', 'embarked', and 'embark_town' contains missing values.

#### **3.2 Let's Drop / Impute the above Columns:**
- Let's drop the 'deck' column from the dataset.
- Let's impute the 'age' and 'fare' columns with the median value.
- Let's impute the 'embarked_town' and 'embarked' columns with the mode value.

In [13]:
# Drop deck column
df.drop('deck', axis=1, inplace=True)

# Impute missing values of age, and fare using median
imputer = SimpleImputer(strategy='median')
df[['age', 'fare']] = imputer.fit_transform(df[['age', 'fare']])

# Impute missing values of embark and embarked_town using mode
imputer = SimpleImputer(strategy='most_frequent')
df[['embark_town', 'embarked']] = imputer.fit_transform(df[['embark_town', 'embarked']])

In [14]:
# check for missing values again
df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

- From the output there is no missing values in the dataset.

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          891 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     891 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  embark_town  891 non-null    object  
 12  alive        891 non-null    object  
 13  alone        891 non-null    bool    
dtypes: bool(2), category(1), float64(2), int64(4), object(5)
memory usage: 79.4+ KB


#### 4. **Encode the Categorical and Object Variables using For Loop & LabelEncoder:** 
- We need to encode the categorical and object variables because the decision tree algorithm does not handle categorical data.
- We can use the `LabelEncoder` class from the `sklearn.preprocessing` module to convert categorical data to numerical data.

In [16]:
# Encode the categorical and object variables using for loop and labelencoder
le = LabelEncoder()
for col in df.select_dtypes(include=['category', 'object']):
    df[col] = le.fit_transform(df[col])

In [17]:
df.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class  who  \
0         0       3    1  22.0      1      0   7.2500         2      2    1   
1         1       1    0  38.0      1      0  71.2833         0      0    2   
2         1       3    0  26.0      0      0   7.9250         2      2    2   
3         1       1    0  35.0      1      0  53.1000         2      0    2   
4         0       3    1  35.0      0      0   8.0500         2      2    1   

   adult_male  embark_town  alive  alone  
0        True            2      0  False  
1       False            0      1  False  
2       False            2      1   True  
3       False            2      1  False  
4        True            2      0   True

#### 5. **Split the Data into Independent 'X' and Dependent 'Y' Variables:**
- We need to split the dataset into independent and dependent variables. The independent variables are the variables that we want to use to make predictions, and the dependent variable is the variable that we want to predict.
- The independent variables are stored in the 'X' variable, and the dependent variable is stored in the 'Y' variable.

In [18]:
# split the data into X and y
X = df.drop(['survived'], axis=1)
y = df['survived']

#### 6. **Split the Data into 80% Training and 20% Testing:**

In [19]:
# split the data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#### 7. **Create and Train the Decision Tree Model:**

In [20]:
# create and train the model with pred
model = DecisionTreeClassifier()
model.fit(X_train, y_train)

DecisionTreeClassifier()

#### 8. **Making Predictions:**

In [21]:
# predict the model
y_pred = model.predict(X_test)

#### 9. **Evaluate the Model:**

In [22]:
# evaluate the model
print(confusion_matrix(y_test, y_pred))
print(classification_report(y_test, y_pred))

[[105   0]
 [  0  74]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00       105
           1       1.00      1.00      1.00        74

    accuracy                           1.00       179
   macro avg       1.00      1.00      1.00       179
weighted avg       1.00      1.00      1.00       179



#### **Observations from the Output:**
| Metric | Class 0 | Class 1 | Overall | Observations |
| --- | --- | --- | --- | --- |
| **Count** | 105 | 74 | 179 | The dataset contains more instances of class 0 than class 1. |
| **Precision** | 1.00 | 1.00 | Weighted Avg: 1.00 | The model has perfect precision for both classes, meaning that every instance that it predicted as a certain class was actually that class. |
| **Recall** | 1.00 | 1.00 | Macro Avg: 1.00 | The model has perfect recall for both classes, meaning that it identified all instances of each class correctly. |
| **F1-Score** | 1.00 | 1.00 | Weighted Avg: 1.00 | The model has a perfect F1 score for both classes, indicating a perfect balance between precision and recall. |
| **Support** | 105 | 74 | 179 | The actual occurrences of each class in the dataset match the count. |
| **Confusion Matrix** | True Positives: 105<br>False Negatives: 0 | False Positives: 0<br>True Negatives: 74 | - | The model made no errors in its predictions. It correctly predicted all instances of both classes. |
| **Accuracy** | - | - | 1.00 | The model has perfect accuracy, meaning it correctly predicted the class of every instance in the dataset. |

Note: In the context of the confusion matrix, "True Positives" and "True Negatives" refer to correct predictions, while "False Positives" and "False Negatives" refer to incorrect predictions.

#### 10. **Save the Model:**

In [23]:
# save the decision tree classifier
from sklearn.tree import export_graphviz
export_graphviz(model, out_file='./saved_models/Decision_tree_01.dot', feature_names=X.columns, filled=True, rounded=True)

# **Decision Tree `Regression` Example in Python**

In [24]:
# 1. Split the data into independent (X) and Dependent (y) variables
X = df.drop(['fare'], axis=1)  # Assuming we want to predict 'fare' for regression
y = df['fare']

# 2. Split the data into 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# 3. Create and train the decision tree model
model = DecisionTreeRegressor()
model.fit(X_train, y_train)

# 4. Making Predictions
y_pred = model.predict(X_test)

# 5. Evaluate the Model
mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error: {mse}")

# 6. Save the Model
from sklearn.tree import export_graphviz
export_graphviz(model, out_file='./saved_models/Decision_tree_regression_01.dot', feature_names=X.columns, filled=True, rounded=True)

Mean Squared Error: 2569.1690617816334
